realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java

```diff
@@ -294,13 +293,18 @@ private boolean categorizeClassElements() {
                 }
 
                 if (variableElement.getAnnotation(Index.class) != null) {
-                    // The field has the @Index annotation. It's only valid for:
-                    // * String
+                    // The field has the @Index annotation. It's only valid for column types:
+                    // STRING, DATE, INTEGER, BOOLEAN
                     String elementTypeCanonicalName = variableElement.asType().toString();
-                    if (elementTypeCanonicalName.equals("java.lang.String")) {
+                    String columnType = RealmProxyClassGenerator.JAVA_TO_COLUMN_TYPES.get(elementTypeCanonicalName);
+                    if (columnType != null && (columnType.equals("ColumnType.STRING") ||
+                            columnType.equals("ColumnType.DATE") ||
+                            columnType.equals("ColumnType.INTEGER") ||
+                            columnType.equals("ColumnType.BOOLEAN"))) {
                         indexedFields.add(variableElement);
                     } else {
-                        Utils.error("@Index is only applicable to String fields - got " + element);
+                        Utils.error("@Index is not applicable to this field " + element +
+                                ", which is a type of " + elementTypeCanonicalName + ".");
```

Comment:
I would probably just use the first line, this is not really adding any value to the error message.